integration-cli/docker_cli_push_test.go

```diff
@@ -134,41 +135,38 @@ func (s *DockerRegistrySuite) TestConcurrentPush(c *testing.T) {
 	var repos []string
 	for _, tag := range []string{"push1", "push2", "push3"} {
 		repo := fmt.Sprintf("%v:%v", imgRepo, tag)
-		buildImageSuccessfully(c, repo, build.WithDockerfile(fmt.Sprintf(`
-	FROM busybox
-	ENTRYPOINT ["/bin/echo"]
-	ENV FOO foo
-	ENV BAR bar
-	CMD echo %s
-`, repo)))
+		buildImageSuccessfully(c, repo, build.WithDockerfile(fmt.Sprintf("FROM busybox\nCMD echo hello from %s\n", repo)))
 		repos = append(repos, repo)
 	}
 
-	// Push tags, in parallel
-	results := make(chan error, len(repos))
+	cleanup := func() {
+		args := append([]string{"rmi", "--force"}, repos...)
+		cli.DockerCmd(c, args...)
+	}
+	defer cleanup()
 
+	// Push tags, in parallel
+	var eg errgroup.Group
 	for _, repo := range repos {
-		go func(repo string) {
+		eg.Go(func() error {
 			result := icmd.RunCommand(dockerBinary, "push", repo)
-			results <- result.Error
-		}(repo)
-	}
-
-	for range repos {
-		err := <-results
-		assert.NilError(c, err, "concurrent push failed with error: %v", err)
+			// check the result, but don't fail immediately (result.Assert)
+			// to prevent the errgroup from not completing.
+			assert.Check(c, result.Equal(icmd.Success))
+			return result.Error
+		})
```

Comment:
For this kind of stuff, I generally prefer to do the assertions outside of the goroutine.

I assume the output here is nicer due to doing `assert.Check(c, result.Equal(icmd.Success))` instead of `assert.NilError(c, err, "concurrent push failed with error: %v", err)`.

I'm also thinking that if `assert.Check(c, result.Equal(icmd.Success))` passes, then there was no error, so we don't need to check both `assert.Check(c, result.Equal(icmd.Success))` and `assert.NilError(c, err, "concurrent push failed with error: %v", err)`

How about something like:
```go
	// Push tags, in parallel
	results := make(chan icmd.Result, len(repos))
	for _, repo := range repos {
		go func(repo string) {
			result := icmd.RunCommand(dockerBinary, "push", repo)
			results <- result
		}(repo)
	}

	for range repos {
		result := <-results
		assert.Check(c, result.Equal(icmd.Success))
	}
```

Or, if

> 'm also thinking that if `assert.Check(c, result.Equal(icmd.Success))` passes, then there was no error, so we don't need to check both `assert.Check(c, result.Equal(icmd.Success))` and `assert.NilError(c, err, "concurrent push failed with error: %v", err)`

isn't true, then:
```go
...
	for range repos {
		result := <-results
		assert.Check(c, result.Equal(icmd.Success))
		assert.NilError(t, result.Error)
	}
```

Reply:
> How about something like:

Heh; fun fact; that's what I originally started with but went a bit back-and-forth; I was looking at "but now we're looping twice" so I was considering if using an error-group would then be cleaner (run them once, and if it doesn't return an error, then it's all good).


I can have a look again if it's possibly cleaner.